main.go

```diff
@@ -0,0 +1,176 @@
+// be updated.
+func toWPTMetadata(testName string, link shared.MetadataLink) {
+	dirpath := path.Dir(testName)
+	testpath := path.Base(testName)
+	createDirIfNeeded(dirpath)
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0755)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
+		metadata := shared.Metadata{
+			Links: []shared.MetadataLink{
+				{
+					Product: link.Product,
+					URL:     link.URL,
+					Results: []shared.MetadataTestResult{{
+						TestPath: testpath,
+					}},
+				},
+			},
+		}
+		writeToFile(metadata, f)
+		return
+	}
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var metadata shared.Metadata
+	err = yaml.Unmarshal(data, &metadata)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// First check whether we have an existing crbug link in this META.yml; if
+	// so we just want to append our testname to the set of tests associated
+	// with it.
+	hasAdded := false
+	for index, existingLink := range metadata.Links {
+		if !link.Product.MatchesProductSpec(existingLink.Product) {
+			continue
+		}
+
+		// Check whether we are already listed in this entry.
+		for _, result := range existingLink.Results {
+			if result.TestPath == testpath {
+				fmt.Println("EXISTING-LINK: for test", testName, "with url", link.URL)
+				return
+			}
+		}
+
+		if existingLink.URL == link.URL {
+			fmt.Println("APPEND-LINK: for test", testName, "with url", link.URL)
+			metadata.Links[index].Results = append(existingLink.Results, shared.MetadataTestResult{TestPath: testpath})
+			hasAdded = true
+			break
+		}
+	}
+
+	// Otherwise, this is a brand new crbug entry in the META.yml, so add a new
+	// link to it.
+	if !hasAdded {
+		fmt.Println("NEW-LINK: for test", testName, "with url", link.URL)
+		newLink := shared.MetadataLink{
+			Product: link.Product,
+			URL:     link.URL,
+			Results: []shared.MetadataTestResult{{
+				TestPath: testpath,
+			}},
+		}
+		metadata.Links = append(metadata.Links, newLink)
+	}
+
+	writeToFile(metadata, f)
+}
+
+// createDirIfNeeded is a helper function to create a directory if it doesn't
+// already exist. It handles recursive directory creation.
+func createDirIfNeeded(dir string) {
+	_, err := os.Stat(dir)
+	if !os.IsNotExist(err) {
+		return
+	}
+	os.MkdirAll(dir, os.ModePerm)
+}
+
+// writeToFile writes out a wpt.fyi shared.Metadata entry to a given file handle.
+//
+// TODO: This seems to cause a lot of existing metadata files to be completely
```

Comment:
Could it be https://github.com/web-platform-tests/wpt.fyi/issues/1957?

This TODO seems serious enough that it should be resolved before merging, otherwise it's hard to vet that no unexpected changes are happening.

Maybe all updates to META.yml files should go through a shared code path, so that different code paths aren't flipping the formatting back and forth between different indentations and stuff?

Reply:
I was able to improve the vet issue by upgrading `"gopkg.in/yaml.v2" to `"gopkg.in/yaml.v3"`. I will link this TODO to  web-platform-tests/wpt.fyi#1957

Reply:
Sounds good!